In [1]:
import os
os.chdir("D:/UIT/HK/HK6/IS403/Đồ án/Source code")
print("Thư mục làm việc hiện tại:", os.getcwd())

Thư mục làm việc hiện tại: D:\UIT\HK\HK6\IS403\Đồ án\Source code


# Import modules

In [2]:
import pandas as pd

In [3]:
import plotly.graph_objects as go
from plotly.subplots import make_subplots

# Load data

In [5]:
tcb_kqkd = pd.read_csv("./dataset/clean_dataset/bank_index/tcb/tcb_kqkd.csv")
tcb_kqkd

Năm  Kỳ  Doanh thu (đồng)  Tăng trưởng doanh thu (%)  \
0   2024   4    15064961000000                   0.042920   
1   2024   3    15168282000000                   0.043733   
2   2024   2    15348981000000                   0.096524   
3   2024   1    14507238000000                   0.056439   
4   2023   4    14444980000000                   0.179178   
5   2023   3    14532716000000                   0.291638   
6   2023   2    13997858000000                   0.325161   
7   2023   1    13732205000000                   0.284815   
8   2022   4    12250039000000                   0.288072   
9   2022   3    11251381000000                   0.233667   
10  2022   2    10563140000000                   0.211169   
11  2022   1    10688076000000                   0.311234   
12  2021   4     9510371000000                   0.240602   

    Lợi nhuận sau thuế của Cổ đông công ty mẹ (đồng)  \
0                                      3371065000000   
1                                      5737505000000   
2                                      6193361000000   
3                                      6220997000000   
4                                      4438909000000   
5                                      4612943000000   
6                                      4455055000000   
7                                      4496895000000   
8                                      3544272000000   
9                                      5297966000000   
10                                     5803525000000   
11                                     5504614000000   
12                                     4592230000000   

    Tăng trưởng lợi nhuận (%)  LN trước thuế  Lợi nhuận thuần  Cố tức đã nhận  
0                   -0.240565  4696032000000    3420111000000        67000000  
1                    0.243784  7214168000000    5793022000000      6406000000  
2                    0.390187  7826509000000    6269968000000     58051000000  
3                    0.383398  7801659000000    6277003000000        37000000  
4                    0.252418  5773070000000    4481615000000     23633000000  
5                   -0.129299  5842790000000    4668705000000      8450000000  
6                   -0.232354  5649010000000    4503449000000        31000000  
7                   -0.183068  5623358000000    4537097000000      2642000000  
8                   -0.228202  4746223000000    3572087000000        30000000  
9                    0.221405  6715079000000    5367540000000      6415000000  
10                   0.231876  7321136000000    5882255000000      2295000000  
11                   0.251979  6785330000000    5614544000000        51000000  
12                   0.161678  6139979000000    4683818000000      4001000000

In [6]:
tcb_cstc = pd.read_csv("./dataset/clean_dataset/bank_index/tcb/tcb_cstc.csv")
tcb_cstc

CP   Năm  Kỳ  Biên lợi nhuận ròng (%)   ROE (%)   ROA (%)  \
0   TCB  2024   4                 0.391902  0.153980  0.023544   
1   TCB  2024   3                 0.642593  0.163577  0.025308   
2   TCB  2024   2                 0.653451  0.160440  0.025072   
3   TCB  2024   1                 0.731921  0.152018  0.024289   
4   TCB  2023   4                 0.584282  0.146945  0.023253   
5   TCB  2023   3                 0.634329  0.142284  0.023307   
6   TCB  2023   2                 0.707727  0.153479  0.025183   
7   TCB  2023   1                 0.688981  0.171774  0.028175   
8   TCB  2022   4                 0.519733  0.195193  0.031789   
9   TCB  2022   3                 0.700283  0.208842  0.034202   
10  TCB  2022   2                 0.744630  0.210537  0.034456   
11  TCB  2022   1                 0.678655  0.210433  0.034342   
12  TCB  2021   4                 0.633883  0.215348  0.035806   

    Vốn hóa (Tỷ đồng)  Số CP lưu hành (Triệu CP)       P/E       P/B  \
0     183686145214000                 7064851739  8.512548  1.263150   
1     166730501040400                 7064851739  7.361654  1.164898   
2     171546276495700                 7045021622  7.989180  1.250613   
3     160626492981600                 7045021622  8.138237  1.176752   
4     166967012441400                 3522510811  9.261686  1.281412   
5     122054999601150                 3522510811  7.124790  0.969395   
6     114310251705000                 3517238514  6.422257  0.942791   
7     111320598968100                 3517238514  5.811075  0.953124   
8     101648193054600                 3517238514  5.038535  0.905180   
9      97779230689200                 3517238514  4.605034  0.898155   
10     88986134404200                 3517238514  4.388570  0.859142   
11    126744024207800                 3510914798  6.614889  1.297252   
12    160097714788800                 3510914798  8.857907  1.736520   

         P/S  P/Cash Flow    EPS (VND)    BVPS (VND)  
0   3.909015     4.323498   477.160049  20583.468751  
1   3.441059     5.262235   812.119661  20259.284053  
2   3.640826     3.093598   879.111709  19470.448263  
3   3.733513     3.418103   883.034479  19375.368072  
4   4.167810    13.834033  1260.154827  36990.442043  
5   3.172653     7.203138  1309.561062  35743.930601  
6   2.977540    10.654725  1266.634316  34472.102053  
7   2.776766    12.109685  1278.530012  33206.575140  
8   2.485160     1.830379  1007.685997  31927.334627  
9   2.348556     3.614172  1506.285678  30952.342460  
10  2.220515     3.058984  1650.023158  29447.986467  
11  3.313036     2.861459  1567.857472  27828.050984  
12  4.318062    12.465108  1307.986740  26259.416507

In [7]:
tcb_cdtc = pd.read_csv("./dataset/clean_dataset/bank_index/tcb/tcb_cdtc.csv")
tcb_cdtc

Năm  Kỳ  TỔNG CỘNG TÀI SẢN (đồng)  TỔNG CỘNG NGUỒN VỐN (đồng)  \
0   2024   4           978798549000000             978798549000000   
1   2024   3           927053345000000             927053345000000   
2   2024   2           908306700000000             908306700000000   
3   2024   1           885653461000000             885653461000000   
4   2023   4           849482012000000             849482012000000   
5   2023   3           781278666000000             781278666000000   
6   2023   2           732470169000000             732470169000000   
7   2023   1           723517800000000             723517800000000   
8   2022   4           699032544000000             699032544000000   
9   2022   3           671353524000000             671353524000000   
10  2022   2           623738504000000             623738504000000   
11  2022   1           615270328000000             615270328000000   
12  2021   4           568811435000000             568811435000000   

    NỢ PHẢI TRẢ (đồng)  VỐN CHỦ SỞ HỮU (đồng)  LỢI ÍCH CỦA CỔ ĐÔNG THIỂU SỐ  
0      830858928000000        147939621000000                 2520466000000  
1      782685363000000        144367982000000                 1437445000000  
2      769755043000000        138551657000000                 1381928000000  
3      747780735000000        137872726000000                 1372839000000  
4      717853947000000        131628065000000                 1316833000000  
5      654148880000000        127129786000000                 1274127000000  
6      610005199000000        122464970000000                 1218365000000  
7      605552384000000        117965416000000                 1169971000000  
8      585607578000000        113424966000000                 1128915000000  
9      561454375000000        109899149000000                 1032378000000  
10     519263345000000        104475159000000                  962804000000  
11     516684338000000         98585990000000                  884074000000  
12     475755598000000         93055837000000                  844828000000

# Perform descriptive statistics

In [8]:
tcb_kqkd.describe()

Năm         Kỳ  Doanh thu (đồng)  Tăng trưởng doanh thu (%)  \
count    13.000000  13.000000      1.300000e+01                  13.000000   
mean   2022.846154   2.615385      1.315848e+13                   0.200396   
std       0.987096   1.192928      2.030567e+12                   0.106119   
min    2021.000000   1.000000      9.510371e+12                   0.042920   
25%    2022.000000   2.000000      1.125138e+13                   0.096524   
50%    2023.000000   3.000000      1.399786e+13                   0.233667   
75%    2024.000000   4.000000      1.453272e+13                   0.288072   
max    2024.000000   4.000000      1.534898e+13                   0.325161   

       Lợi nhuận sau thuế của Cổ đông công ty mẹ (đồng)  \
count                                      1.300000e+01   
mean                                       4.943795e+12   
std                                        9.291765e+11   
min                                        3.371065e+12   
25%                                        4.455055e+12   
50%                                        4.612943e+12   
75%                                        5.737505e+12   
max                                        6.220997e+12   

       Tăng trưởng lợi nhuận (%)  LN trước thuế  Lợi nhuận thuần  \
count                  13.000000   1.300000e+01     1.300000e+01   
mean                    0.086403   6.318026e+12     5.005478e+12   
std                     0.246969   1.050719e+12     9.415170e+11   
min                    -0.240565   4.696032e+12     3.420111e+12   
25%                    -0.183068   5.649010e+12     4.503449e+12   
50%                     0.221405   6.139979e+12     4.683818e+12   
75%                     0.251979   7.214168e+12     5.793022e+12   
max                     0.390187   7.826509e+12     6.277003e+12   

       Cố tức đã nhận  
count    1.300000e+01  
mean     8.623769e+09  
std      1.618511e+10  
min      3.000000e+07  
25%      5.100000e+07  
50%      2.642000e+09  
75%      6.415000e+09  
max      5.805100e+10

In [9]:
tcb_cdtc.describe()

Năm         Kỳ  TỔNG CỘNG TÀI SẢN (đồng)  \
count    13.000000  13.000000              1.300000e+01   
mean   2022.846154   2.615385              7.665205e+14   
std       0.987096   1.192928              1.326654e+14   
min    2021.000000   1.000000              5.688114e+14   
25%    2022.000000   2.000000              6.713535e+14   
50%    2023.000000   3.000000              7.324702e+14   
75%    2024.000000   4.000000              8.856535e+14   
max    2024.000000   4.000000              9.787985e+14   

       TỔNG CỘNG NGUỒN VỐN (đồng)  NỢ PHẢI TRẢ (đồng)  VỐN CHỦ SỞ HỮU (đồng)  \
count                1.300000e+01        1.300000e+01           1.300000e+01   
mean                 7.665205e+14        6.444158e+14           1.221047e+14   
std                  1.326654e+14        1.151889e+14           1.767182e+13   
min                  5.688114e+14        4.757556e+14           9.305584e+13   
25%                  6.713535e+14        5.614544e+14           1.098991e+14   
50%                  7.324702e+14        6.100052e+14           1.224650e+14   
75%                  8.856535e+14        7.477807e+14           1.378727e+14   
max                  9.787985e+14        8.308589e+14           1.479396e+14   

       LỢI ÍCH CỦA CỔ ĐÔNG THIỂU SỐ  
count                  1.300000e+01  
mean                   1.272690e+12  
std                    4.213118e+11  
min                    8.448280e+11  
25%                    1.032378e+12  
50%                    1.218365e+12  
75%                    1.372839e+12  
max                    2.520466e+12

In [11]:
tcb_cstc.describe()

Năm         Kỳ  Biên lợi nhuận ròng (%)    ROE (%)    ROA (%)  \
count    13.000000  13.000000                13.000000  13.000000  13.000000   
mean   2022.846154   2.615385                 0.639413   0.175758   0.028364   
std       0.987096   1.192928                 0.096127   0.027892   0.004966   
min    2021.000000   1.000000                 0.391902   0.142284   0.023253   
25%    2022.000000   2.000000                 0.633883   0.153479   0.024289   
50%    2023.000000   3.000000                 0.653451   0.163577   0.025308   
75%    2024.000000   4.000000                 0.700283   0.208842   0.034202   
max    2024.000000   4.000000                 0.744630   0.215348   0.035806   

       Vốn hóa (Tỷ đồng)  Số CP lưu hành (Triệu CP)        P/E        P/B  \
count       1.300000e+01               1.300000e+01  13.000000  13.000000   
mean        1.363460e+14               4.605599e+09   6.932797   1.130645   
std         3.270737e+13               1.699580e+09   1.620570   0.245667   
min         8.898613e+13               3.510915e+09   4.388570   0.859142   
25%         1.113206e+14               3.517239e+09   5.811075   0.942791   
50%         1.267440e+14               3.517239e+09   7.124790   1.164898   
75%         1.667305e+14               7.045022e+09   8.138237   1.263150   
max         1.836861e+14               7.064852e+09   9.261686   1.736520   

             P/S  P/Cash Flow    EPS (VND)    BVPS (VND)  
count  13.000000    13.000000    13.000000     13.000000  
mean    3.269578     6.440701  1169.703474  28193.596155  
std     0.683194     4.292759   338.006259   6459.251112  
min     2.220515     1.830379   477.160049  19375.368072  
25%     2.776766     3.093598   883.034479  20583.468751  
50%     3.313036     4.323498  1266.634316  29447.986467  
75%     3.733513    10.654725  1309.561062  33206.575140  
max     4.318062    13.834033  1650.023158  36990.442043

# Visualize

## Kết quả kinh doanh

In [12]:
tcb_kqkd["Tháng"] = tcb_kqkd["Kỳ"] * 3
tcb_kqkd

Năm  Kỳ  Doanh thu (đồng)  Tăng trưởng doanh thu (%)  \
0   2024   4    15064961000000                   0.042920   
1   2024   3    15168282000000                   0.043733   
2   2024   2    15348981000000                   0.096524   
3   2024   1    14507238000000                   0.056439   
4   2023   4    14444980000000                   0.179178   
5   2023   3    14532716000000                   0.291638   
6   2023   2    13997858000000                   0.325161   
7   2023   1    13732205000000                   0.284815   
8   2022   4    12250039000000                   0.288072   
9   2022   3    11251381000000                   0.233667   
10  2022   2    10563140000000                   0.211169   
11  2022   1    10688076000000                   0.311234   
12  2021   4     9510371000000                   0.240602   

    Lợi nhuận sau thuế của Cổ đông công ty mẹ (đồng)  \
0                                      3371065000000   
1                                      5737505000000   
2                                      6193361000000   
3                                      6220997000000   
4                                      4438909000000   
5                                      4612943000000   
6                                      4455055000000   
7                                      4496895000000   
8                                      3544272000000   
9                                      5297966000000   
10                                     5803525000000   
11                                     5504614000000   
12                                     4592230000000   

    Tăng trưởng lợi nhuận (%)  LN trước thuế  Lợi nhuận thuần  Cố tức đã nhận  \
0                   -0.240565  4696032000000    3420111000000        67000000   
1                    0.243784  7214168000000    5793022000000      6406000000   
2                    0.390187  7826509000000    6269968000000     58051000000   
3                    0.383398  7801659000000    6277003000000        37000000   
4                    0.252418  5773070000000    4481615000000     23633000000   
5                   -0.129299  5842790000000    4668705000000      8450000000   
6                   -0.232354  5649010000000    4503449000000        31000000   
7                   -0.183068  5623358000000    4537097000000      2642000000   
8                   -0.228202  4746223000000    3572087000000        30000000   
9                    0.221405  6715079000000    5367540000000      6415000000   
10                   0.231876  7321136000000    5882255000000      2295000000   
11                   0.251979  6785330000000    5614544000000        51000000   
12                   0.161678  6139979000000    4683818000000      4001000000   

    Tháng  
0      12  
1       9  
2       6  
3       3  
4      12  
5       9  
6       6  
7       3  
8      12  
9       9  
10      6  
11      3  
12     12

In [13]:
fig = go.Figure()

fig.add_trace(go.Scatter(
    x=tcb_kqkd["Năm"].astype(str) + "-" + tcb_kqkd["Tháng"].astype(str),
    y=tcb_kqkd["Doanh thu (đồng)"],
    mode="lines+markers",
    name="Doanh thu (đồng)"
))

fig.update_layout(
    title="Doanh thu (đồng) theo thời gian",
    xaxis_title="Năm - Kỳ",
    yaxis_title="Doanh thu (đồng)"
)

fig.show()


Theo biểu đồ, ta thấy:
- Từ đầu năm 2022 - 2025, doanh thu có xu hướng tăng và đạt đỉnh ở quý 2 năm 2024. Giai đoạn tăng mạnh nhất từ quý 2 2022 - quý 1 năm 2023.
- Sau quý 2 năm 2024, doanh thu có xu hướng giảm nhẹ.

In [14]:
fig = go.Figure()

fig.add_trace(go.Bar(
    x=tcb_kqkd["Năm"].astype(str) + "-" + tcb_kqkd["Tháng"].astype(str),
    y=tcb_kqkd["Tăng trưởng doanh thu (%)"],
    marker_color=["green" if val > 0 else "red" for val in tcb_kqkd["Tăng trưởng doanh thu (%)"]]
))

fig.update_layout(
    title="Tăng trưởng doanh thu (%) theo thời gian",
    xaxis_title="Năm - Kỳ",
    yaxis_title="Tăng trưởng doanh thu (%)"
)

fig.show()


Theo biểu đồ tăng trưởng doanh thu, ta thấy:
- Từ đầu năm 2022 - 2024, tăng trưởng doanh thu luôn dương. Trong đó, quý 2 năm 2023 tăng trưởng doanh thu đạt đỉnh.
- Trong năm 2022 - quý 2 2023, tăng trưởng doanh thu cao nhưng ổn định.
- Từ quý 2 năm 2023, tăng trưởng doanh thu có xu hướng giảm. Và đạt thấp nhất từ quý 4 năm 2024.

In [15]:
fig = go.Figure()

fig.add_trace(go.Bar(
    x=tcb_kqkd["Năm"].astype(str) + "-" + tcb_kqkd["Tháng"].astype(str),
    y=tcb_kqkd["Tăng trưởng lợi nhuận (%)"],
    marker_color=["green" if val > 0 else "red" for val in tcb_kqkd["Tăng trưởng lợi nhuận (%)"]]
))

fig.update_layout(
    title="Tăng trưởng lợi nhuận (%) theo thời gian",
    xaxis_title="Năm - Kỳ",
    yaxis_title="Tăng trưởng lợi nhuận (%)"
)

fig.show()


Theo biểu đồ tăng trưởng lợi nhuận, ta thấy:
- Từ năm 2022 đến quý 3 năm 2022, tăng trưởng lợi nhuận dương, trong đó quý 1 năm 2022 tăng trưởng đạt đỉnh.
- Tăng trưởng lợi nhuận từ năm 2023 có xu hướng giảm, trong đó từ cuối năm 2022 đến quý 3 năm 2023 tăng trưởng âm.
- Tăng trưởng lợi nhuận đạt đáy vào cuối năm 2024.

In [16]:
fig = go.Figure()

fig.add_trace(go.Scatter(
    x=tcb_kqkd["Năm"].astype(str) + "-" + tcb_kqkd["Tháng"].astype(str),
    y=tcb_kqkd["LN trước thuế"],
    mode="lines+markers",
    name="Lợi nhuận trước thuế (đồng)"
))
fig.add_trace(go.Scatter(
    x=tcb_kqkd["Năm"].astype(str) + "-" + tcb_kqkd["Tháng"].astype(str),
    y=tcb_kqkd["Lợi nhuận thuần"],
    mode="lines+markers",
    name="Lợi nhuận thuần (đồng)"
))

fig.update_layout(
    title="Lợi nhuận trước thuế và lợi nhuận thuần theo thời gian",
    xaxis_title="Năm - Kỳ",
    yaxis_title="Giá trị (đồng)"
)

fig.show()


Theo biểu đồ, ta thấy:
- Năm 2022 cả lợi nhuận trước thuế và sau thuế đều có xu hướng giảm và đạt đáy vào cuối năm 2022.
- Năm 2023, lợi nhuận trước thuế và sau thuế ổn định.
- Đầu năm 2024, lợi nhuận trước thuế và sau thuế tăng mạnh và đạt đỉnh. Sau đó từ quý 3, lợi nhuận bắt đầu giảm mạnh.

In [17]:
fig.add_trace(go.Bar(
    x=tcb_kqkd["Năm"].astype(str) + "-" + tcb_kqkd["Tháng"].astype(str),
    y=tcb_kqkd["Cố tức đã nhận"],
    name="Cổ tức đã nhận"
))

fig.update_layout(
    title="Cổ tức đã nhận theo thời gian",
    xaxis_title="Năm - Kỳ",
    yaxis_title="Cổ tức đã nhận",
    xaxis_tickangle=-45
)

fig.show()


Theo biểu đồ, ta thấy cổ tức không ổn định.

## Cân đối kế toán

In [18]:
tcb_cdtc["Tháng"] = tcb_cdtc["Kỳ"] * 3
tcb_cdtc

Năm  Kỳ  TỔNG CỘNG TÀI SẢN (đồng)  TỔNG CỘNG NGUỒN VỐN (đồng)  \
0   2024   4           978798549000000             978798549000000   
1   2024   3           927053345000000             927053345000000   
2   2024   2           908306700000000             908306700000000   
3   2024   1           885653461000000             885653461000000   
4   2023   4           849482012000000             849482012000000   
5   2023   3           781278666000000             781278666000000   
6   2023   2           732470169000000             732470169000000   
7   2023   1           723517800000000             723517800000000   
8   2022   4           699032544000000             699032544000000   
9   2022   3           671353524000000             671353524000000   
10  2022   2           623738504000000             623738504000000   
11  2022   1           615270328000000             615270328000000   
12  2021   4           568811435000000             568811435000000   

    NỢ PHẢI TRẢ (đồng)  VỐN CHỦ SỞ HỮU (đồng)  LỢI ÍCH CỦA CỔ ĐÔNG THIỂU SỐ  \
0      830858928000000        147939621000000                 2520466000000   
1      782685363000000        144367982000000                 1437445000000   
2      769755043000000        138551657000000                 1381928000000   
3      747780735000000        137872726000000                 1372839000000   
4      717853947000000        131628065000000                 1316833000000   
5      654148880000000        127129786000000                 1274127000000   
6      610005199000000        122464970000000                 1218365000000   
7      605552384000000        117965416000000                 1169971000000   
8      585607578000000        113424966000000                 1128915000000   
9      561454375000000        109899149000000                 1032378000000   
10     519263345000000        104475159000000                  962804000000   
11     516684338000000         98585990000000                  884074000000   
12     475755598000000         93055837000000                  844828000000   

    Tháng  
0      12  
1       9  
2       6  
3       3  
4      12  
5       9  
6       6  
7       3  
8      12  
9       9  
10      6  
11      3  
12     12

In [19]:
fig = go.Figure()

fig.add_trace(go.Scatter(
    x=tcb_cdtc["Năm"].astype(str) + "-" + tcb_cdtc["Tháng"].astype(str),
    y=tcb_cdtc["TỔNG CỘNG TÀI SẢN (đồng)"],
    mode="lines+markers",
    name="Tổng cộng tài sản (đồng)"
))
fig.add_trace(go.Scatter(
    x=tcb_cdtc["Năm"].astype(str) + "-" + tcb_cdtc["Tháng"].astype(str),
    y=tcb_cdtc["NỢ PHẢI TRẢ (đồng)"],
    mode="lines+markers",
    name="Nợ phải trả (đồng)"
))

fig.update_layout(
    title="Tổng cộng tài sản và nợ phải trả theo thời gian",
    xaxis_title="Năm - Kỳ",
    yaxis_title="Giá trị (đồng)"
)

fig.show()

Theo biểu đồ, ta thấy:
- Tổng cộng tài sản và nợ phải trả có xu hướng tăng dần.
- Vốn chủ sở hữu tăng dần theo thời gian.
- Lợi ích cổ đông thiểu số tăng dần theo thời gian. Đặc biệt là vào cuối năm 2024, lợi ích cổ đông thiểu số tăng mạnh.

In [20]:
fig = go.Figure()

fig.add_trace(go.Scatter(
    x=tcb_cdtc["Năm"].astype(str) + "-" + tcb_cdtc["Tháng"].astype(str),
    y=tcb_cdtc["VỐN CHỦ SỞ HỮU (đồng)"],
    mode="lines+markers",
    name="Vốn chủ sở hữu (đồng)"
))

fig.update_layout(
    title="Vốn chủ sở hữu theo thời gian",
    xaxis_title="Năm - Kỳ",
    yaxis_title="Vốn chủ sở hữu (đồng)"
)

fig.show()


In [21]:
fig = go.Figure()

fig.add_trace(go.Scatter(
    x=tcb_cdtc["Năm"].astype(str) + "-" + tcb_cdtc["Tháng"].astype(str),
    y=tcb_cdtc["LỢI ÍCH CỦA CỔ ĐÔNG THIỂU SỐ"],
    mode="lines+markers",
    name="Lợi ích của cổ dông thiểu số"
))

fig.update_layout(
    title="Lợi ích của cổ dông thiểu số theo thời gian",
    xaxis_title="Năm - Kỳ",
    yaxis_title="Lợi ích của cổ dông thiểu số"
)

fig.show()

## Chỉ số tài chính

In [22]:
tcb_cstc["Tháng"] = tcb_cstc["Kỳ"] * 3
tcb_cstc

CP   Năm  Kỳ  Biên lợi nhuận ròng (%)   ROE (%)   ROA (%)  \
0   TCB  2024   4                 0.391902  0.153980  0.023544   
1   TCB  2024   3                 0.642593  0.163577  0.025308   
2   TCB  2024   2                 0.653451  0.160440  0.025072   
3   TCB  2024   1                 0.731921  0.152018  0.024289   
4   TCB  2023   4                 0.584282  0.146945  0.023253   
5   TCB  2023   3                 0.634329  0.142284  0.023307   
6   TCB  2023   2                 0.707727  0.153479  0.025183   
7   TCB  2023   1                 0.688981  0.171774  0.028175   
8   TCB  2022   4                 0.519733  0.195193  0.031789   
9   TCB  2022   3                 0.700283  0.208842  0.034202   
10  TCB  2022   2                 0.744630  0.210537  0.034456   
11  TCB  2022   1                 0.678655  0.210433  0.034342   
12  TCB  2021   4                 0.633883  0.215348  0.035806   

    Vốn hóa (Tỷ đồng)  Số CP lưu hành (Triệu CP)       P/E       P/B  \
0     183686145214000                 7064851739  8.512548  1.263150   
1     166730501040400                 7064851739  7.361654  1.164898   
2     171546276495700                 7045021622  7.989180  1.250613   
3     160626492981600                 7045021622  8.138237  1.176752   
4     166967012441400                 3522510811  9.261686  1.281412   
5     122054999601150                 3522510811  7.124790  0.969395   
6     114310251705000                 3517238514  6.422257  0.942791   
7     111320598968100                 3517238514  5.811075  0.953124   
8     101648193054600                 3517238514  5.038535  0.905180   
9      97779230689200                 3517238514  4.605034  0.898155   
10     88986134404200                 3517238514  4.388570  0.859142   
11    126744024207800                 3510914798  6.614889  1.297252   
12    160097714788800                 3510914798  8.857907  1.736520   

         P/S  P/Cash Flow    EPS (VND)    BVPS (VND)  Tháng  
0   3.909015     4.323498   477.160049  20583.468751     12  
1   3.441059     5.262235   812.119661  20259.284053      9  
2   3.640826     3.093598   879.111709  19470.448263      6  
3   3.733513     3.418103   883.034479  19375.368072      3  
4   4.167810    13.834033  1260.154827  36990.442043     12  
5   3.172653     7.203138  1309.561062  35743.930601      9  
6   2.977540    10.654725  1266.634316  34472.102053      6  
7   2.776766    12.109685  1278.530012  33206.575140      3  
8   2.485160     1.830379  1007.685997  31927.334627     12  
9   2.348556     3.614172  1506.285678  30952.342460      9  
10  2.220515     3.058984  1650.023158  29447.986467      6  
11  3.313036     2.861459  1567.857472  27828.050984      3  
12  4.318062    12.465108  1307.986740  26259.416507     12

In [23]:
fig = go.Figure()

fig.add_trace(go.Scatter(
    x=tcb_cstc["Năm"].astype(str) + "-" + tcb_cstc["Tháng"].astype(str),
    y=tcb_cstc["ROE (%)"],
    mode="lines+markers",
    name="ROE (%)",
    yaxis="y1"
))

fig.add_trace(go.Scatter(
    x=tcb_cstc["Năm"].astype(str) + "-" + tcb_cstc["Tháng"].astype(str),
    y=tcb_cstc["ROA (%)"],
    mode="lines+markers",
    name="ROA (%)",
    yaxis="y2"
))

fig.update_layout(
    title="ROE và ROA theo thời gian",
    xaxis_title="Năm - Kỳ",
    yaxis=dict(title="ROE (%)", tickfont=dict(color="#1f77b4")),
    yaxis2=dict(title="ROA (%)", overlaying="y", side="right", tickfont=dict(color="#ff7f0e"))
)

fig.show()

Đối với ROE:
- Quý 1 2022 - Quý 3 2023: ROE có xu hướng giảm đáng kể từ khoảng 21% xuống còn khoảng 14%.
- Quý 4 2023 - Quý 4 2024: ROE có dấu hiệu phục hồi và tăng trở lại.

Đối với ROA:
- Quý 1 2022 - Quý 4 2023: ROA có xu hướng giảm tương tự như ROE, từ khoảng 3.3% xuống còn 2.3%.
- Quý 1 2024 - Quý 4 2024: ROA phục hồi và tăng nhẹ. Nhưng và cuối năm, ROE quay đầu giảm trở lại.

So sánh ROE và ROA:
- Cả ROE và ROA đều có xu hướng giảm mạnh như nhau trong giai đoạn từ đầu năm 2022 đến đầu năm 2024.
- Đến giai đoạn cuối, cả hai chỉ số đều có dấu hiệu phục hồi, tuy nhiên mức độ phục hồi của ROE rõ rệt hơn so với ROA.
- ROE luôn cao hơn ROA, điều này là hợp lý vì ROE đo lường lợi nhuận trên vốn chủ sở hữu, trong khi ROA đo lường lợi nhuận trên tổng tài sản (bao gồm cả nợ). Sự khác biệt giữa ROE và ROA phản ánh tác động của đòn bẩy tài chính. Tuy vậy, ROE và ROA có tỉ lệ tương quan cao với nhau.

In [24]:
fig = go.Figure(data=[go.Bar(
    x=tcb_cstc["Năm"].astype(str) + "-" + tcb_cstc["Tháng"].astype(str),
    y=tcb_cstc["Biên lợi nhuận ròng (%)"],
    name="Biên lợi nhuận ròng (%)"
)])

fig.update_layout(
    title="Biên lợi nhuận ròng (%) theo thời gian",
    xaxis_title="Năm - Kỳ",
    yaxis_title="Biên lợi nhuận ròng (%)",
    xaxis_tickangle=-45
)

fig.show()


Theo biểu đồ, ta thấy:
- Năm 2022: Biên lợi nhuận ở mức vừa, sau đó tăng cao nhất ở quý 2, rồi giảm mạnh ở cuối năm.
- Biên lợi nhuận có tính chu kỳ theo năm. Từ mức cao ở đầu năm và giảm dần về cuối năm.
- Quý 4/2024: Biên lợi nhuận ròng giảm xuống mức thấp nhất trong toàn bộ chu kỳ được thể hiện trên biểu đồ (khoảng 39%).

In [25]:
fig = go.Figure()

fig.add_trace(go.Scatter(
    x=tcb_cstc["Năm"].astype(str) + "-" + tcb_cstc["Tháng"].astype(str),
    y=tcb_cstc["Vốn hóa (Tỷ đồng)"],
    mode="lines+markers",
    name="Vốn hóa (tỷ đồng)"
))

fig.update_layout(
    title="Vốn hóa (tỷ đồng) theo thời gian",
    xaxis_title="Năm - Kỳ",
    yaxis_title="Vốn hóa (tỷ đồng)"
)

fig.show()

Theo biểu đồ, ta thấy:
- Quý 1 năm 2022 - Quý 2 năm 2022: Vốn hóa giảm mạnh từ khoảng 126 nghìn tỷ đồng xuống mức thấp nhất là khoảng 89 nghìn tỷ đồng. Đây là giai đoạn giảm sâu nhất trong toàn bộ chu kỳ.
- Từ sau quý 2 năm 2022: Vốn hoá có xu hướng tăng. Trong đó giai đoạn từ quý 2 2022 - quý 3 2023, vốn hoá tăng đều từ khoảng 89 nghìn tỷ lên 122 nghìn tỷ. Từ sau quý 3 2023 - quý 4 2024, vốn hoá tăng mạnh và có biến động mạnh.

In [26]:
fig = go.Figure()

fig.add_trace(go.Bar(
    x=tcb_cstc["Năm"].astype(str) + "-" + tcb_cstc["Tháng"].astype(str),
    y=tcb_cstc["Số CP lưu hành (Triệu CP)"],
    name="Số cổ phiếu lưu hành (triệu cổ phiếu)"
))

fig.update_layout(
    title="Số cổ phiếu lưu hành (triệu cổ phiếu) theo thời gian",
    xaxis_title="Năm - Kỳ",
    yaxis_title="Số cổ phiếu lưu hành (triệu cổ phiếu)",
    xaxis_tickangle=-45
)

fig.show()


Từ biểu đồ, ta thấy số cổ phiếu lưu hành có 1 lần điều chỉnh vào quý 1 năm 2024.

In [27]:
fig = make_subplots(specs=[[{"secondary_y": True}]])

fig.add_trace(
    go.Bar(
        x=tcb_cstc["Năm"].astype(str) + "-" + tcb_cstc["Tháng"].astype(str),
        y=tcb_cstc["BVPS (VND)"],
        name="BVPS"
    ),
    secondary_y=False
)
fig.add_trace(
    go.Scatter(
        x=tcb_cstc["Năm"].astype(str) + "-" + tcb_cstc["Tháng"].astype(str),
        y=tcb_cstc["EPS (VND)"],
        mode="lines+markers",
        name="EPS"
    ),
    secondary_y=True
)

fig.update_layout(
    title="BVPS và EPS theo thời gian",
    xaxis_title="Năm - Kỳ",
    yaxis_title="BVPS",
    yaxis2=dict(title="EPS", overlaying="y", side="right")
)

fig.show()

Về BVPS:
- Quý 1/2022 - Quý 4/2023: BVPS có xu hướng tăng trưởng khá ổn định trong giai đoạn này. Từ khoảng 28 nghìn đồng vào đầu năm 2022, BVPS tăng dần qua các quý và đạt mức cao nhất khoảng 37 nghìn đồng vào cuối năm 2023.
- Quý 1/2024 - Quý 4/2025: BVPS giảm xuống mức 19 nghìn đồng, sau đó có dấu hiệu đi ngang và duy trì ở mức tương đương, dao động nhẹ quanh mốc 20 nghìn đồng.

Về EPS:
- Năm 2022: EPS có sự biến động lớn trong giai đoạn này. Ban đầu ở mức khá cao, EPS giảm mạnh xuống mức thấp nhất vào cuối năm 2022.
- Năm 2023: EPS cho thấy sự phục hồi rõ rệt và có xu hướng ổn định hơn, dao động quanh mức 1250-1300.
- Từ đầu năm 2024, EPS bắt đầu có xu hướng giảm dần và giảm mạnh vào cuối năm 2024.

So sánh BVPS và EPS:
- BVPS cho thấy sự tăng trưởng ổn định trong phần lớn giai đoạn, cho thấy giá trị tài sản ròng trên mỗi cổ phiếu của công ty đang tăng lên.
- EPS lại có sự biến động mạnh hơn. Giai đoạn đầu có sự sụt giảm đáng kể, sau đó phục hồi nhưng lại suy giảm mạnh vào cuối chu kỳ. Điều này cho thấy lợi nhuận trên mỗi cổ phiếu của công ty không ổn định và có dấu hiệu suy yếu trong giai đoạn gần đây.
- Sự khác biệt trong xu hướng của BVPS và EPS có thể cho thấy công ty đang tích lũy tài sản nhưng hiệu quả sinh lời trên tài sản đó lại giảm sút.

In [29]:
fig = make_subplots(
    rows=3,
    cols=1,
    shared_xaxes=True,
    vertical_spacing=0.03,
    subplot_titles=("P/E theo thời gian", "P/B theo thời gian", "P/S theo thời gian")
)

fig.add_trace(
    go.Scatter(
        x=tcb_cstc["Năm"].astype(str) + "-" + tcb_cstc["Tháng"].astype(str),
        y=tcb_cstc["P/E"], name="P/E"
    ),
    row=1,
    col=1
)
fig.add_trace(
    go.Scatter(
        x=tcb_cstc["Năm"].astype(str) + "-" + tcb_cstc["Tháng"].astype(str),
        y=tcb_cstc["P/B"], name="P/B"
    ),
    row=2,
    col=1
)
fig.add_trace(
    go.Scatter(
        x=tcb_cstc["Năm"].astype(str) + "-" + tcb_cstc["Tháng"].astype(str),
        y=tcb_cstc["P/S"], name="P/S"
    ),
    row=3,
    col=1
)

fig.update_layout(height=800, width=800, title_text="P/E, P/B, và P/S Ratios")
fig.update_xaxes(tickangle=-45, tickfont=dict(size=10), row=3, col=1)

fig.show()

P/E Ratio Over Time:
- Quý 1/2022 - Quý 2/2022: Tỷ lệ P/E giảm từ khoảng 6.5 xuống mức thấp nhất là khoảng 4.5 vào quý 2/2022.
- Quý 2/2022 - Quý 4/2023: Tỷ lệ P/E có xu hướng tăng lên, đạt đỉnh khoảng 9.2 vào cuối năm 2023.
- Năm 2024: Tỷ lệ P/E có sự biến động nhẹ và giảm nhẹ, nhưng vào cuối năm tỷ lệ có sự tăng trở lại.

P/B Ratio Over Time:
- Quý 1/2022 - Quý 2/2022: Tỷ lệ P/B có sự biến động mạnh, giảm mạnh từ khoảng 1.3 xuống dưới 0.9.
- Quý 2/2023 - Quý 3/2023: Tỷ lệ P/B có xu hướng tăng nhẹ.
- Quý 3/2024 - Quý 4/2024: Tỷ lệ P/B có có biến động mạnh, từ cuối năm 2023, tỷ lệ tăng mạnh, sau đó biến động lên xuống quanh mức 1.15 - 1.3.

P/S Ratio Over Time:
- Quý 1/2022 - Quý 2/2022: Tỷ lệ P/B có sự biến động mạnh, giảm mạnh từ khoảng 3.3 xuống dưới 2.2.
- Quý 2/2022 - Quý 4/2023: Tỷ lệ P/S có xu hướng tăng trưởng khá ổn định, đạt đỉnh khoảng 4.2.
- Năm 2024: Tỷ lệ P/S có sự điều chỉnh giảm xuống nhưng vẫn duy trì trên mức 3.5.

In [30]:
fig = go.Figure(data=[go.Bar(
    x=tcb_cstc["Năm"].astype(str) + "-" + tcb_cstc["Tháng"].astype(str),
    y=tcb_cstc["P/Cash Flow"],
    marker_color=["red" if value < 0 else "green" for value in tcb_cstc["P/Cash Flow"]]
)])

fig.update_layout(
    title="P/Cash Flow theo thời gian",
    xaxis_title="Năm - Tháng",
    yaxis_title="P/Cash Flow",
    xaxis_tickangle=-45
)

fig.show()

Dựa trên biểu đồ, ta thấy:
- Năm 2022: Tỷ lệ P/Cash Flow có xu hướng giảm mạnh. Đạt mức thấp nhất là khoảng 1.8.
- Năm 2023: Tỷ lệ P/Cash Flow có sự tăng trưởng vượt bậc và duy trì quanh mức cao. Và đạt đỉnh vào cuối năm 2023 khoảng 14.
- Năm 2024: Sau khi đạt đỉnh, tỷ lệ P/Cash Flow giảm mạnh xuống khoảng 3.4 và 3.1, phục hồi nhẹ lên 5.3 và cuối cùng giảm trở lại khoảng 4.4 vào cuối kỳ.